Jellyfin.Server.Implementations/Devices/DeviceManager.cs

```diff
@@ -0,0 +1,240 @@
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading.Tasks;
+using Jellyfin.Data.Entities;
+using Jellyfin.Data.Entities.Security;
+using Jellyfin.Data.Enums;
+using Jellyfin.Data.Events;
+using Jellyfin.Data.Queries;
+using MediaBrowser.Controller.Devices;
+using MediaBrowser.Controller.Library;
+using MediaBrowser.Model.Devices;
+using MediaBrowser.Model.Querying;
+using MediaBrowser.Model.Session;
+using Microsoft.EntityFrameworkCore;
+
+namespace Jellyfin.Server.Implementations.Devices
+{
+    /// <summary>
+    /// Manages the creation, updating, and retrieval of devices.
+    /// </summary>
+    public class DeviceManager : IDeviceManager
+    {
+        private readonly JellyfinDbProvider _dbProvider;
+        private readonly IUserManager _userManager;
+        private readonly ConcurrentDictionary<string, ClientCapabilities> _capabilitiesMap = new ();
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DeviceManager"/> class.
+        /// </summary>
+        /// <param name="dbProvider">The database provider.</param>
+        /// <param name="userManager">The user manager.</param>
+        public DeviceManager(JellyfinDbProvider dbProvider, IUserManager userManager)
+        {
+            _dbProvider = dbProvider;
+            _userManager = userManager;
+        }
+
+        /// <inheritdoc />
+        public event EventHandler<GenericEventArgs<Tuple<string, DeviceOptions>>>? DeviceOptionsUpdated;
+
+        /// <inheritdoc />
+        public void SaveCapabilities(string deviceId, ClientCapabilities capabilities)
+        {
+            _capabilitiesMap[deviceId] = capabilities;
```

Comment:
I'm not strong about this but:


I already raised my concerns about having users stored in memory rather than querying the database due to the problems it posed during the migrations in #6049 and some random concurrency errors that users had.

With this pattern we're essentially locking us down with no federation, as multiple instances running of a single database will have "cached" in memory the information. Changing data in one side will trigger concurrency errors in the other one.

I know we're far away from that, but imo we should lay new code that has the best foundation possible for every single thing we might need one day. 

If we want to keep doing this, I would at least add a comment to all the places where we follow this pattern as a reminder.

Reply:
The way we handle device capabilities is going to change for 11, and if we ever do decide to pursue federation, it will be post 11 anyways. As for the issues with concurrency, the only ones I've seen are unrelated to the users being kept in memory. While I agree that it's suboptimal, until we redo our session/auth code, they need to be cached since they are accessed so often. For 11, we'll likely do some form of JWT and can do permissions/validation on the data in the token, at which point we can stop caching users.